torch/csrc/jit/frontend/exit_transforms.cpp

```diff
@@ -505,6 +529,260 @@ struct ExitTransformer {
+
+  DepthFirstGraphNodeIterator it(graph);
+  Node* node = it.next();
+
+  while (node) {
+    if (node->kind() == prim::Enter) {
+      exit_node_stack.emplace_back(node);
+    } else if (node->kind() == prim::Exit) {
+      // exit_node_stack should not be empty.
+      TORCH_INTERNAL_ASSERT(!exit_node_stack.empty());
+      // The input to this Exit node should be the same as that of the Enter
+      // node on the top of the exit_node_stack.
+      TORCH_INTERNAL_ASSERT(exit_node_stack.back()->input(0) == node->input(0));
+      // Record the pair.
+      enter_exit_pairs.emplace_back(exit_node_stack.back(), node);
+      exit_node_stack.pop_back();
+    }
+
+    node = it.next();
+  }
+
+  // The stack should not be empty; an Exit should have been found for every
+  // Enter.
+  TORCH_INTERNAL_ASSERT(exit_node_stack.empty());
+
+  // Now, add a With block for each Enter-Exit pair. The innermost pairs were
+  // found first, so they will be converted first.
+  for (auto& pair : enter_exit_pairs) {
+    Node* enter = pair.first;
+    Node* exit = pair.second;
+
+    auto* with = graph->create(prim::With, /*num_outputs=*/0);
+    auto* body_block = with->addBlock();
+    auto* exit_block = with->addBlock();
+
+    // Insert the With after the Enter.
+    Node* cur = enter->next();
+    Node* insert_point = body_block->param_node();
+
+    // Move all of the nodes between the Enter and Exit into the body block.
+    while (cur != exit) {
+      auto* next = cur->next();
+      cur->moveAfter(insert_point);
+      insert_point = insert_point->next();
+      cur = next;
+    }
+
+    // Move the Exit node into the exit block.
+    exit->moveAfter(exit_block->param_node());
+    with->insertAfter(enter);
+  }
+}
+
+// Removes prim::With nodes from a graph. More specifically, it transforms
+// IR that looks like this:
+//
+//   %a = prim::Enter(%b)
+//   = prim::With()
+//     block0():
+//       <code>
+//     -> ()
+//     block1():
+//       %c = prim::Exit(%b)
+//      ->()
+//
+// to this:
+//   %a = prim::Enter(%b)
+//   <code>
+//   %c = prim::Exit(%b)
+//
+static void convertWithBlocksToEnterExitNodes(std::shared_ptr<Graph>& graph) {
+  // First, find all With blocks to avoid iterator invalidation issues when
+  // moving nodes around later.
+  std::vector<Node*> with_nodes;
+
+  DepthFirstGraphNodeIterator it(graph);
+  Node* node = it.next();
+
+  while (node) {
+    if (node->kind() == prim::With) {
+      with_nodes.emplace_back(node);
+    }
+    node = it.next();
+  }
+
+  // For each With node:
+  for (auto& node : with_nodes) {
+    auto* body_block = node->blocks().at(0);
+    auto* exit_block = node->blocks().at(1);
+
+    std::vector<Node*> to_append;
+
+    // Record all nodes that need to be appended after the Enter that precedes
+    // the With block to avoid iterator invalidation issues later when moving
+    // nodes around.
+    for (auto it = body_block->nodes().begin(); it != body_block->nodes().end();
```

Comment:
i forget to do this too but a nit is just to use `(for auto node: block->nodes())` if you're not doing any invalidation while you're iterating